pkg/registry/service/allocator/etcd/etcd.go

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	etcderr "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors/etcd"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service/allocator"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/tools"
+)
+
+// Etcd exposes a service.Allocator that is backed by etcd.
+// TODO: allow multiple allocations to be tried at once
+// TODO: subdivide the keyspace to reduce conflicts
+// TODO: investigate issuing a CAS without reading first
+type Etcd struct {
+	lock sync.Mutex
+
+	alloc  allocator.Snapshottable
+	helper tools.EtcdHelper
+	last   string
+}
+
+// Etcd implements allocator.Interface and service.IPRegistry
+var _ allocator.Interface = &Etcd{}
+var _ service.IPRegistry = &Etcd{}
+
+const baseKey = "/serviceipallocations"
```

Comment:
----- Original Message -----
> > +// TODO: allow multiple allocations to be tried at once
> > +// TODO: subdivide the keyspace to reduce conflicts
> > +// TODO: investigate issuing a CAS without reading first
> > +type Etcd struct {
> > +	lock sync.Mutex
> > +
> > +	alloc  allocator.Snapshottable
> > +	helper tools.EtcdHelper
> > +	last   string
> > +}
> > +
> > +// Etcd implements allocator.Interface and service.IPRegistry
> > +var _ allocator.Interface = &Etcd{}
> > +var _ service.IPRegistry = &Etcd{}
> > +
> > +const baseKey = "/serviceipallocations"
> 
> if I understand the model, this creates a single "file" in etcd for
> serviceipallocations.  If we have a port allocator (we will soon) then we
> duplicate this whole directory tree and change a string constant?
> 
> Wouldn't we be better to have a top-level /master/ranges dir into which we
> would write a file called "service-ips" and another called "service-ports" ?
> 

Maybe.  I think eventually you might want to shard to multiple ranges, so absent knowing whether we needed to deal with both single ranges and sharded ranges and per namespace ranges I just tried to get the basics.  I prefer having a clear keyspace that deals with ranges (unless you want a recursive delete of a certain keyspace to delete the ranges?).